backend/lucene/src/main/java/org/hibernate/search/backend/lucene/types/predicate/impl/LuceneNumericTermsPredicate.java

```diff
@@ -0,0 +1,107 @@
+ */
+package org.hibernate.search.backend.lucene.types.predicate.impl;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.hibernate.search.backend.lucene.search.impl.AbstractLuceneCodecAwareSearchValueFieldQueryElementFactory;
+import org.hibernate.search.backend.lucene.search.impl.LuceneSearchContext;
+import org.hibernate.search.backend.lucene.search.impl.LuceneSearchValueFieldContext;
+import org.hibernate.search.backend.lucene.search.predicate.impl.AbstractLuceneLeafSingleFieldPredicate;
+import org.hibernate.search.backend.lucene.types.codec.impl.AbstractLuceneNumericFieldCodec;
+import org.hibernate.search.engine.search.common.ValueConvert;
+import org.hibernate.search.engine.search.predicate.SearchPredicate;
+import org.hibernate.search.engine.search.predicate.spi.TermsPredicateBuilder;
+
+import org.apache.lucene.search.BooleanClause;
+import org.apache.lucene.search.BooleanQuery;
+import org.apache.lucene.search.Query;
+
+public class LuceneNumericTermsPredicate extends AbstractLuceneLeafSingleFieldPredicate {
+
+	private LuceneNumericTermsPredicate(Builder builder) {
+		super( builder );
+	}
+
+	public static class Factory<F, E extends Number>
+			extends
+			AbstractLuceneCodecAwareSearchValueFieldQueryElementFactory<TermsPredicateBuilder, F, AbstractLuceneNumericFieldCodec<F, E>> {
+		public Factory(AbstractLuceneNumericFieldCodec<F, E> codec) {
+			super( codec );
+		}
+
+		@Override
+		public Builder<F, E> create(LuceneSearchContext searchContext, LuceneSearchValueFieldContext<F> field) {
+			return new Builder<>( codec, searchContext, field );
+		}
+	}
+
+	private static class Builder<F, E extends Number> extends AbstractBuilder<F> implements TermsPredicateBuilder {
+		private final AbstractLuceneNumericFieldCodec<F, E> codec;
+
+		private E term;
+		private List<E> terms;
+		private boolean allMatch;
+
+		private Builder(AbstractLuceneNumericFieldCodec<F, E> codec, LuceneSearchContext searchContext,
+				LuceneSearchValueFieldContext<F> field) {
+			super( searchContext, field );
+			this.codec = codec;
+		}
+
+		@Override
+		public void matchingAny(Object term, Object... terms) {
+			allMatch = false;
+			fillTerms( term, terms );
+		}
+
+		@Override
+		public void matchingAll(Object term, Object... terms) {
+			allMatch = true;
+			fillTerms( term, terms );
+		}
+
+		@Override
+		public SearchPredicate build() {
+			return new LuceneNumericTermsPredicate( this );
+		}
+
+		@Override
+		protected Query buildQuery() {
+			if ( term != null ) {
+				return codec.getDomain().createExactQuery( absoluteFieldPath, term );
+			}
+
+			if ( !allMatch ) {
+				return codec.getDomain().createSetQuery( absoluteFieldPath, terms );
+			}
+
+			BooleanQuery.Builder builder = new BooleanQuery.Builder();
+			for ( E termItem : terms ) {
+				Query query = codec.getDomain().createExactQuery( absoluteFieldPath, termItem );
+				builder.add( query, BooleanClause.Occur.MUST );
+			}
```

Comment:
That's not great: there is a (relatively low) limit to how many clauses you can add to a boolean query, and one of the purposes of this new `terms` predicate is to overcome that limitation.

But from what I can see there isn't any alternative, unless we want to roll out our own query... 

Ok, let's go with this. Make sure to test queries with many terms (`> 1024`) for both cases, though (`matchingAll` and `matchingAny`). I suspect you will get an exception for `matchingAll` and will have to add a test exception through `TckBackendFeatures`.

Reply:
Queries with many terms wouldn't be supported in any case, see my comment above. At most I would check and validate the size of the terms. Going to do that.

Reply:
Thinking better about that, since we have the same issue ( of validate the max clause count ), maybe it is better to move it to a different issue : https://hibernate.atlassian.net/browse/HSEARCH-4205.

Reply:
Queries with many terms (significantly more than 1024) should be supported for `matchAny` (when using `TermInSetQuery`). 

I don't remember if you already did, but please make sure you test at least that?

Reply:
I'm adding now...